### PR TITLE
Fix product import for Product page v2.

### LINKED ISSUE #28169: BO - An exception is displayed in the product page V2 when importing a product without supplier
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

I tried to import a product without a supplier, an exception is displayed
[product.csv](https://github.com/PrestaShop/PrestaShop/files/8443619/product.csv)
image
before, I have this exception
image

### Expected behavior

_No response_

### Steps to reproduce

1. Go to BO > Advanced Parameters > Import product csv without supplier
[product.csv](https://github.com/PrestaShop/PrestaShop/files/8443625/product.csv)
2. Go to BO > Catalog > Products V2
3. Edit the imported product

In the Product V1 => ok

### PrestaShop version(s) where the bug happened

8.x

### PHP version(s) where the bug happened

7.4, 8.0

### If your bug is related to a module, specify its name and its version

_No response_

### PR DESCRIPTION
**Pull request by MeKeyCool**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 8.0.x
| Description?      | Product page v2 display an error after a product import.
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #28169.
| Related PRs       | Na.
| How to test?      | Cf. #28169 
| Possible impacts? | Product creation.

> :question: for @PrestaShop/prestashop-core-developers , the solution to create a new Product to use `Product::getDynamicProductType()` function according Product state isn't completely satisfying. But with @atomiix we didn't find an easy solution that won't imply deep refactoring. If anyone as a better solution please share :)

In order to fix product import for Product page v2, we 
- Remove default empty supplier creation during import.
    > Import calls legacy controller that creates some weird data with 0 ids (associative tables). Inducing wrong behaviors when trying to get consistent data with product page v2.
- Fix getProductType for `ProductType::TYPE_UNDEFINED`
    > `ProductType::TYPE_UNDEFINED` is kept for legacy shop compatibility (existing databases, import old datas, ...) but a new function as been implemented to manage data consistency : `Product::getDynamicProductType()`. We searched a way to build the ProductType from CQRS Query request while product isn't instantiated yet and it's data are not satisfying new required structure.

### CODE REVIEW

**Review by matks — COMMENTED** [review]
> Import calls legacy controller that creates some weird data with 0 ids (associative tables). Inducing wrong behaviors when trying to get consistent data with product page v2.

PrestaShop 8.0.0 allows users to use product page v1 or v2. So what will happen for users of product page v1?

File: src/Adapter/Product/Repository/ProductRepository.php [hunk 1]
```diff
@@ -301,6 +301,11 @@ public function getProductType(ProductId $productId): ProductType
             ));
         }
 
+        if ($result['product_type'] === ProductType::TYPE_UNDEFINED) {
```
Comment (jolelievre) [inline src/Adapter/Product/Repository/ProductRepository.php]: That's a good initiative to make this method more resilient to absent data in the DB! I was afraid that maybe we were actually relying on the fact that this exception needs to be thrown in some cases but it doesn't seem to be the case.

To validate this new behavior and make sure we don't include some regression later we should enforce this with a test. This could be tested in behat test by creating a Product with an undefined type maybe in an integration test

File: controllers/admin/AdminImportController.php [hunk 2]
```diff
@@ -2104,17 +2104,14 @@ protected function productImportOne($info, $default_language_id, $id_lang, $forc
                 $id_product_supplier = (int) ProductSupplier::getIdByProductAndSupplier((int) $product->id, 0, (int) $product->id_supplier);
                 if ($id_product_supplier) {
                     $product_supplier = new ProductSupplier($id_product_supplier);
-                } else {
-                    $product_supplier = new ProductSupplier();
+                    $product_supplier->id_product = (int) $product->id;
```
Comment (jolelievre) [inline controllers/admin/AdminImportController.php]: I'm not sure about this modification, it means that importing a new supplier that doesn't exist will never work anymore so in most cases import of product supplier will remain empty..

**Review by jolelievre — CHANGES_REQUESTED** [review]
Thanks @MeKeyCool 

One test should be added, and I'm not convinced about the modification in import, seems like it induces a regression

File: controllers/admin/AdminImportController.php (hunk 2 above)
Comment (atomiix) [reply controllers/admin/AdminImportController.php]: @jolelievre that's actually what happens when you create a product with product page v1 when you don't specify a supplier, that's why we made that choice.

Comment (jolelievre) [reply controllers/admin/AdminImportController.php]: Hmm but I'm not sure it is the same use case 🤔 In product page we are editing a product which can have an existing ProductSupplier, in this case we are importing a product, so most of the time a new product so the chances are high that the ProductSupplier will never exist for this couple of `id_product/id_supplier` meaning if it's not present the association is never created

I have close to zero knowledge about the import feature, but it really feels like in this case the ProductSupplier can only be updated when it exists and can never created be created by import

Comment (atomiix) [reply controllers/admin/AdminImportController.php]: @jolelievre I'm not sure to understand why it's different?
In product page you can also create a product without defining a product supplier so to me it's like importing without specifying a product supplier?
Also, if I remember well, you can have a product supplier during import, you just have to specify which one

Comment (jolelievre) [reply controllers/admin/AdminImportController.php]: I think the main thing to check here is if `!empty($product->id_supplier)` to make sure we don't create an association to a non existing supplier

But if the association doesn't exist (and that's the purpose of `ProductSupplier::getIdByProductAndSupplier`) we still need to create it (and not update it)

Comment (atomiix) [reply controllers/admin/AdminImportController.php]: @jolelievre you're right, I didn't get it right 😅

**Comment by MeKeyCool** [comment]
> > Import calls legacy controller that creates some weird data with 0 ids (associative tables). Inducing wrong behaviors when trying to get consistent data with product page v2.
> 
> PrestaShop 8.0.0 allows users to use product page v1 or v2. So what will happen for users of product page v1?

@matks , do you mean that you suspect I didn't test enough ?
Would you be satisfied f I add some behat tests as suggested by @jolelievre ?
If you have some specific scenario in your mind, please share ^^

File: controllers/admin/AdminImportController.php (hunk 2 above)
Comment (MeKeyCool) [reply controllers/admin/AdminImportController.php]: I am a bit lost with this issue so I may be out of the subject.
But @jolelievre , isn't the test you are asking for done at line 2103 ?

```php
if (!$validateOnly && !empty($product->id) && property_exists($product, 'supplier_reference')) {
```

File: src/Adapter/Product/Repository/ProductRepository.php (hunk 1 above)
Comment (MeKeyCool) [reply src/Adapter/Product/Repository/ProductRepository.php]: After some investigations, I found those tests : https://github.com/PrestaShop/PrestaShop/blob/develop/tests/Integration/Behaviour/Features/Scenario/Product/legacy_product_type.feature
Isn't it enough ? What test is lacking for you ?

Comment (jolelievre) [reply src/Adapter/Product/Repository/ProductRepository.php]: This test was already present before your fix, and yet a bug was happening when we requested the product's suppliers using the `GetProductSupplierOptions` query Which means this use case was not tested (trying to ftehc the suppliers while the type is undefined)

So our test coverage didn't handle this buggy use case, meaning the use case wasn't present and should be added, and you're right it should probably be tested in the feature you mentioned

Basically, you can add a test that tries and get the suppliers, if you remove your fix it should fail but with the fix it should be fine now

File: controllers/admin/AdminImportController.php (hunk 2 above)
Comment (jolelievre) [reply controllers/admin/AdminImportController.php]: Hmm, I'm not talking about the reference value but about the actual `$product->id_supplier` which stores the relation between the product and its default supplier

What I mean in this comment is:
- with your modification, you only create the `ProductSupplier` when `$id_product_supplier` is not empty
- `$id_product_supplier` is initialized with `ProductSupplier::getIdByProductAndSupplier` which searches for a relationship between the product and `$product->id_supplier`, meaning it will only return a value if the supplier is already associated to the product (ProductSupplier represents the association table between product and supplier)
- so if the relationship isn't found (because it's not in DB) you don't create the ProductSupplier, meaning you don't create the relationship
- so an inexistent relationship will never be created
- so with you modify you can only modify/update an existing relationship
- TLDR: if the relationship doesn't exist yet it's never created so the import is now unable to create a new relationship (it was able to do so with the previous code)

So that's why I'm blocking this modification because it introduces a regression

Comment (jolelievre) [reply controllers/admin/AdminImportController.php]: Now besides this modification which is not appropriate I discussed with @atomiix who worked with you on this bug fix and it seems like there was a bug because of this because a ProductSupplier instance was created but the id_supplier was 0 which is not accepted in the product page v2 because it tries to get a non-existent supplier (indeed no supplier can have an ID 0) But the problem comes from the fact that we created a ProductSuplier relationship with id_supplier being 0

This should be prevented, and to do so we need to check that `$product->id_supplier` is different from 0 before we try and create the ProductSupplier (and we don't even need to look for a relationship with `ProductSupplier::getIdByProductAndSupplier` since it doesn't make sense)

Comment (jolelievre) [reply controllers/admin/AdminImportController.php]: TLDR: you should add `&& !empty($product->id_supplier)` on Line 2103, currently the test checks the `supplier_reference` value not the ID itself

File: src/Adapter/Product/Repository/ProductRepository.php (hunk 1 above)
Comment (MeKeyCool) [reply src/Adapter/Product/Repository/ProductRepository.php]: Just to be precise, this fix is not fixing initial problem, when we investigated initial issue, we found two different problems.

I shared a behat test but I am not well experienced with it so I expect you correct me if I did anything wrong ^^'

File: tests/Integration/Behaviour/Features/Context/LegacyProductFeatureContext.php [hunk 3]
```diff
@@ -986,6 +987,22 @@ public function productPriceAfterReduction($productName, $priceWithReduction)
         }
     }
 
+    /**
+     * @Then product :productName as to be editable
```
Comment (FabienPapet) [inline tests/Integration/Behaviour/Features/Context/LegacyProductFeatureContext.php]: `has* to be editable`

File: src/Adapter/Product/Repository/ProductRepository.php [hunk 4]
```diff
@@ -310,12 +310,14 @@ public function getProductType(ProductId $productId): ProductType
             ));
         }
 
-        if (!empty($result['product_type'])) {
-            return new ProductType($result['product_type']);
+        // Older products that were created before product page v2, might have no type, so we determine it dynamically
+        // According Phpstan, "Strict comparison using === between mixed and '' will always evaluate to false." so we use == instead of ===
+        if (empty($result['product_type']) || ($result['product_type'] == ProductType::TYPE_UNDEFINED)) {
```
Comment (atomiix) [inline src/Adapter/Product/Repository/ProductRepository.php]: I think you should annotate the `$result` variable instead of using `==`:
```php
/** @var array<string, string> $result */
$result = $this->connection->createQueryBuilder()
...
```

Comment (MeKeyCool) [reply src/Adapter/Product/Repository/ProductRepository.php]: Lol, in fact I was completely out of subject, the point was that `ProductType::TYPE_UNDEFINED = ''` (empty string) so `empty($result['product_type'])` already test this case :sweat_smile:

File: tests/Integration/Behaviour/Features/Context/LegacyProductFeatureContext.php [hunk 5]
```diff
@@ -986,6 +987,22 @@ public function productPriceAfterReduction($productName, $priceWithReduction)
         }
     }
 
+    /**
+     * @Then product :productName has to be editable
+     */
+    public function productAsToBeEditable($productName)
```
Comment (atomiix) [inline tests/Integration/Behaviour/Features/Context/LegacyProductFeatureContext.php]:
```suggestion
     * @Then product :productName should be editable
     */
    public function productShouldBeEditable($productName)
```

File: tests/Integration/Behaviour/Features/Scenario/Product/legacy_product_type.feature [hunk 6]
```diff
@@ -64,3 +64,11 @@ Feature: Legacy products have consistent product type through dynamic checking (
...
     And product "pack_product" persisted type should be pack
     And product "pack_product" dynamic type should be pack
+
+  Scenario: I create a product with undefined type using legacy methods, this product as to be editable
+    Given there is a product in the catalog named "Undefined Product" with a price of 15.0 and 100 items in stock
+    Then there is a product "undefined_product" with name "Undefined Product"
+    And product "undefined_product" type should be standard
+    And product "undefined_product" persisted type should be undefined
+    And product "undefined_product" dynamic type should be standard
+    And product "Undefined Product" has to be editable
```
Comment (atomiix) [inline tests/Integration/Behaviour/Features/Scenario/Product/legacy_product_type.feature]: I know it's not what has been done with the tests just above but it make more sense because `there is a product "undefined_product" with name "Undefined Product"` doesn't check something, it sets something.
```suggestion
    And there is a product "undefined_product" with name "Undefined Product"
    Then product "undefined_product" type should be standard
    And product "undefined_product" persisted type should be undefined
    And product "undefined_product" dynamic type should be standard
    And product "Undefined Product" should be editable
```

File: src/Adapter/Product/Repository/ProductRepository.php [hunk 7]
```diff
@@ -309,7 +309,7 @@ public function getProductType(ProductId $productId): ProductType
                 $productId->getValue()
             ));
         }
-
+
```
Comment (atomiix) [inline src/Adapter/Product/Repository/ProductRepository.php]:
```suggestion

```

**Review by atomiix — CHANGES_REQUESTED** [review]
There's at least the method `productAsToBeEditable` that needs to be updated and I think `productShouldBeEditable` is a better name to be consistent with the other `*should be*` methods

**Review by jolelievre — APPROVED** [review]
Thanks @MeKeyCool

**Review by atomiix — APPROVED** [review]

**Comment by sallemiines** [comment]
I will check this pr

**Comment by sallemiines** [comment]
Hello @MeKeyCool 

- The issu is solved 
- I run the automated tests **OK**
- In the BO > Advanced settings > Import csv product without supplier
- Go to BO > Catalog > V2 Products
- Edit the imported product
- The product is edited successfully
- Pass an order with an imported product
- Delete the product
- In the BO > Advanced settings > Import csv product with supplier 
- Go to BO > Catalog > V2 Products
- Edit the imported product
- The product is edited successfully

https://user-images.githubusercontent.com/31663369/189140716-81015ad2-00c3-40c0-8a87-4534e7539272.mp4


Thank you !

**Comment by matthieu-rolland** [comment]
good job @MeKeyCool and @sLorenzini !
